admob/src/android/admob_android.cc

```diff
@@ -441,33 +453,151 @@ JNIEnv* GetJNI() {
 
 jobject GetActivity() { return (g_app) ? g_app->activity() : g_activity; }
 
+void AdmobInternal::CompleteLoadAdFuture(
+    FutureCallbackData<LoadAdResult>* callback_data, int error_code,
+    const std::string& error_message,
+    const LoadAdResultInternal& load_ad_result_internal) {
+  callback_data->future_data->future_impl.CompleteWithResult(
+      callback_data->future_handle, static_cast<int>(error_code),
+      error_message.c_str(), LoadAdResult(load_ad_result_internal));
+  // This method is responsible for disposing of the callback data struct.
+  delete callback_data;
+}
+
 static void CompleteAdFutureCallback(JNIEnv* env, jclass clazz, jlong data_ptr,
                                      jint error_code, jstring error_message) {
   if (data_ptr == 0) return;  // test call only
 
   const char* error_msg = env->GetStringUTFChars(error_message, nullptr);
 
-  firebase::admob::FutureCallbackData* callback_data =
-      reinterpret_cast<firebase::admob::FutureCallbackData*>(data_ptr);
+  firebase::admob::FutureCallbackData<void>* callback_data =
+      reinterpret_cast<firebase::admob::FutureCallbackData<void>*>(data_ptr);
 
-  CompleteFuture(static_cast<int>(error_code), error_msg,
-                 callback_data->future_handle, callback_data->future_data);
+  callback_data->future_data->future_impl.Complete(
+      callback_data->future_handle, static_cast<int>(error_code), error_msg);
 
   env->ReleaseStringUTFChars(error_message, error_msg);
 
   // This method is responsible for disposing of the callback data struct.
   delete callback_data;
 }
 
+void CompleteLoadAdCallback(FutureCallbackData<LoadAdResult>* callback_data,
+                            jobject j_load_ad_error, AdMobError error_code,
+                            const std::string& error_message) {
+  FIREBASE_ASSERT(callback_data);
+
+  std::string future_error_message;
+  LoadAdResultInternal load_ad_result_internal;
+  AdResultInternal* adr = &(load_ad_result_internal.ad_result_internal);
+
+  adr->j_ad_error = j_load_ad_error;
+  adr->is_successful = true;  // assume until proven otherwise.
+  adr->is_wrapper_error = false;
+  adr->code = error_code;
+
+  // Further result configuration is based on success/failure.
+  if (j_load_ad_error != nullptr) {
+    // The Android SDK returned an error.  Use the j_ad_error object
+    // to populate a LoadAdResult with the error specifics.
+    adr->is_successful = false;
+  } else if (adr->code != kAdMobErrorNone) {
+    // C++ SDK Android AdMob Wrapper encountered an error.
+    adr->is_wrapper_error = true;
+    adr->is_successful = false;
+    adr->message = error_message;
+    adr->domain = "SDK";
+    adr->to_string = std::string("Internal error: ") + adr->message;
+    future_error_message = adr->message;
+  }
+
+  // Invoke a friend of LoadAdResult to have it invoke the LoadAdResult
+  // protected constructor with the LoadAdResultInternal data.
+  AdmobInternal::CompleteLoadAdFuture(
```

Comment:
Why doesn't CompleteLoadAdFuture just take an AdResultInternal instead of wrapping it in a LoadAdResultInternal?

Reply:
Fixed.